### ISSUE TITLE
Cold starts double to quadruple when layer is included

### ISSUE THREAD

**Issue by ianwremmel** [aws-observability/aws-otel-lambda#228]
**Is your feature request related to a problem? Please describe.**
I'm just getting started with OpenTelemetry and just the act of adding the ADOT NodeJS layer to my functions makes the simplest ping function from 1-2 to 4-5 seconds and my remix app's cold start goes from 8 seconds to more than 30 (I've got the function set to timeout at 30). 

Is there any configuration that I can do to improve that or is that just the way it has to be?

**Describe the solution you'd like**
Minimal impact on cold start times.

**Describe alternatives you've considered**
- use standard xray libs
- use otel with non-AWS providers and send directly from application code

**Comment by NathanielRN** [comment]
Hi @ianwremmel, thanks for raising this issue!

We've actually gotten issues on things like this before. A user on the OpenTelemetry CNCF Slack channel did a useful deep dive into investigating [cold start impacts of Lambda with the ADOT NodeJS Lambda Layer](https://cloud-native.slack.com/archives/C01U8U07Y4D/p1641168416049200).

Unfortunately it seems that the best solution is to bump up the memory capacity of your Lambda which does improve Cold Start time. If your Lambda is being used a lot this should prove to be a good idea.

Let me know if that helps!

**Comment by ianwremmel** [comment]
Gotcha, thanks!

Could you share that post somehow? I don't seem to be able to access that Slack.

**Comment by NathanielRN** [comment]
> Could you share that post somehow? I don't seem to be able to access that Slack.

<img width="499" alt="image" src="https://user-images.githubusercontent.com/23139455/156638495-b3bc64a8-0866-430c-8826-779d63db2014.png">

<img width="497" alt="image" src="https://user-images.githubusercontent.com/23139455/156638606-459f4727-ad0e-49f5-b9b0-feff8e9be65c.png">
<img width="498" alt="image" src="https://user-images.githubusercontent.com/23139455/156638687-707173fc-713d-47ed-bbf1-4bbfe41a796b.png">

Here is [the Discussion on OTel JS that the user linked](https://github.com/open-telemetry/opentelemetry-js/discussions/2698).

**Comment by ianwremmel** [comment]
Thanks! yea, it looks I'm in kind of the same boat. I didn't see the kind of speedup he did by increasing memory, but even if I had, that would be way too expensive. I don't think there's any way I can use AWS otel, given these numbers, but maybe there's still a chance to use otel instrumentation and ship directly to third-party receivers (e.g., honeycomb).

**Comment by NathanielRN** [comment]
One improvement you might try to reduce the cold start time is to use ESBuild instead of CommonJS build for your Node JS app on Lambda. ESBuild is nicer because it does operations like tree-shaking to seriously reduce the size of imported libraries which reduces memory and improves performance.

There are discussions upstream on OpenTelemetry JS Contrib to improve support for ESBuild because it requires more careful configuration. You might find the comment thread at https://github.com/open-telemetry/opentelemetry-js-contrib/issues/647#issuecomment-982258484 helpful towards this goal and improve your Node App's memory impact in a way that would make ADOT Lambda layers usable for your purpose 🙂

The solution utilizes esbuild's "external" option and a custom plugin. "external" is an array of packages you tell esbuild to not bundle, and the plugin picks up those external packages and installs them into a raw node_modules directory when building. OpenTelemetry needs to monkey patch methods, so you can't bundle the libraries it needs to instrument (e.g. `http`, `express`, `graphql`, etc.), but you can instrument everything else.

**Comment by ianwremmel** [comment]
I'm already using rollup to do pretty aggressive treeshaking and I've tried putting all the otel packages on my dependencies layer. Unfortunately, as soon as I include the AWS Open Telemetry layer, cold boot times skyrocket.

I've made a few attempts at switch my lambdas over to `mjs`, but things go wrong with dependencies every time.

Note that the ping lambda I've been talking about above is about 40k after being processed by rollup.

**Comment by ianwremmel** [comment]
came back to poke at this a little bit more and it seems like most of the overhead comes from adding `http:` to protocols in `opentelemetry.yml`. If I stick to only `grpc:`, then cold starts are only about 10% worse rather than 400% worse.

**Comment by ianwremmel** [comment]
Another interesting detail. This line from [the docs site](https://aws-otel.github.io/docs/getting-started/js-sdk/trace-manual-instr#sending-traces-to-aws-x-ray), when used with a remix app, seems to mostly prevent the function from returning. From what I can tell, initialization takes more than 30 seconds. (For reasons not entirely clear to me, I couldn't get the lambda timeout to be more than 30 seconds, but that might be a sleep-deprivation issue on my part). 

`const tracerProvider = new NodeTracerProvider(tracerConfig);`

**Comment by ianwremmel** [comment]
Actually, that's the wrong line. I think it's _this_ line:

```
import {NodeTracerProvider} from '@opentelemetry/node';
```

Apparently there's some side effect somewhere in `@opentelemetry/node` that completely tanks lambda initialization when Remix is involved.

**Comment by ianwremmel** [comment]
> For reasons not entirely clear to me, I couldn't get the lambda timeout to be more than 30 seconds, but that might be a sleep-deprivation issue on my part

It's the API gateway timeout that has hard limit at 30 seconds. Looking at the corresponding xray trace, it looks like the lambda takes 1.2 minutes to initialize when both remix and `@opentelemetry/node` are present.

Admittedly, remix is only the most obvious package in this lambda. There are any number of frontend packages that I could imagine might not play well with otel; the bundle file is 3MB.

**Comment by ianwremmel** [comment]
After a bunch of refactoring to get `AWSXRayIdGenerator` and `AWSXRayPropagator` working, both of the problems I mentioned above seem to have gone away: the http receiver and `@opentelementr/node` + remix are no longer making initialization times skyrocket. I don't really know what meaningfully changed.

**Comment by willarmiros** [comment]
> After a bunch of refactoring to get AWSXRayIdGenerator and AWSXRayPropagator working

@ianwremmel why did you need to refactor to get these components to work? They should work out of the box with the Lambda layer instrumentation

**Comment by ianwremmel** [comment]
@willarmiros I was having a whole lot of trouble getting the layer to actually collect my spans and send them anywhere. 

First, the layer doesn't configure a few things that one would otherwise expect it to configure:
- it only configures otel to talk to the collector via http, not grpc
- it does not configure a `ContextProvider`

a side effect is that following, say, [the aws otel javascript tracing guide](https://aws-otel.github.io/docs/getting-started/js-sdk/metric-manual-instr) has no effect, because there's already been a tracer configured by the layer. Since otel doesn't provide a way to get access to the original `tracerProvider,` there's no way to add grpc after the fact. (it took me a while, but once I actually found the aws and otel repos for the layer, I was able to configure things using the hooks provided by the upstream otel layer to address the issues I mentioned above.)

Second, again because there's no way to get the `tracerProvider` after initialization, I had to piggyback on one of the config hooks so I could attach the `traceProvider` to `global` so that I could call `global.traceProvider.foceFlush()` at the end of each of my lambda functions.

Finally, there have been a lot of API changes the majority of [aws-otel.github.io](https://aws-otel.github.io/) was written, so they're all slightly incorrect. 

In any case, after working through those issues, my original coldstart issue and the super high duration on my Remix lambdas went away. 

This is what my final set of config overrides looks like
```ts
// Copied from
// https://github.com/open-telemetry/opentelemetry-lambda/blob/0a83149fe2f23a7dab64df6108cfa35f18cc2ae5/nodejs/packages/layer/src/wrapper.ts#L42-L50
declare global {
  // in case of downstream configuring span processors etc
  function configureTracerProvider(tracerProvider: NodeTracerProvider): void;

  function configureTracer(defaultConfig: NodeTracerConfig): NodeTracerConfig;

  function configureSdkRegistration(
    defaultSdkRegistration: SDKRegistrationConfig
  ): SDKRegistrationConfig;

  function configureLambdaInstrumentation(
    config: AwsLambdaInstrumentationConfig
  ): AwsLambdaInstrumentationConfig;
}

// And this needs to be defined so that we can flush at the end of each
// invocation.
declare global {
  // var seems to be the only typescript keyword that works here.
  // eslint-disable-next-line no-var
  var tracerProvider: NodeTracerProvider;
}

// Remove some noise from the service name
const serviceName = process.env.AWS_LAMBDA_FUNCTION_NAME?.replace(
  my-stack-name',
  ''
)?.split('-')?.[0];

// This produces the config passed to new NodeTracerProvider()
global.configureTracer = ({resource, ...config}) => {
  diag.debug('Telemetry: generating tracer config');

  if (serviceName) {
    const r = new Resource({
      [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]:
        process.env.STAGE_NAME ?? 'development',
      [SemanticResourceAttributes.FAAS_INSTANCE]:
        process.env.AWS_LAMBDA_LOG_STREAM_NAME ?? '',
      [SemanticResourceAttributes.FAAS_MAX_MEMORY]:
        process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE ?? '',
      [SemanticResourceAttributes.SERVICE_NAME]: serviceName,
    });
    resource = resource ? resource.merge(r) : r;
  }

  return {
    ...config,
    idGenerator: new AWSXRayIdGenerator(),
    resource,
  };
};

// This gets applied to the tracerProvider
global.configureTracerProvider = (tracerProvider) => {
  diag.debug('Telemetry: configuring tracer provider');
  const exporter = new OTLPTraceExporter({
    url: 'grpc://localhost:4317',
  });

  tracerProvider.addSpanProcessor(new BatchSpanProcessor(exporter));

  global.tracerProvider = tracerProvider;
};

// This produces the config passed to tracerProvider.register();
global.configureSdkRegistration = (config) => {
  diag.debug('Telemetry: generating SDK registration config');
  const contextManager = new AsyncHooksContextManager();
  contextManager.enable();
  return {
    ...config,
    contextManager
  };
};
```

**Comment by ianwremmel** [comment]
So, it wasn't really `AWSXRayIdGenerator` and `AWSXRayPropagator` that weren't working, it was just that the layer doesn't quite work out of the box. When I wrote the original comment, I thought the issue was with the AWS pieces, but it mostly the fact that the layer doesn't flush on its own.

**Comment by kmihaltsov** [comment]
Having similar performance issue. Nodejs Lambda with 1536mb: cold start goes from 500ms with x-ray sdk to 1700ms with this layer and default configuration from aws docs.

What makes me crazy is that aws documentation suggests to use this OTEL layer over x-ray while still having such issues.

**Comment by a-h** [comment]
Similar situation with Go.

I was testing out following the advice to switch to Open Telemetry, spent a few hours on troubleshooting etc., only to find that the Lambda layer adds 200ms(!) to cold start times compared to using the X-Ray SDK.

Disappointed to see the advice to look at Open Telemetry, without any warning of the performance penalty.

The only reason I did was because I thought I could cut out some of the cold start time (the Go AWS X-Ray has about 6MB of dependency bloat for me).

With 1GB of RAM, I'm seeing init duration go from around 100ms, to 300ms.

I don't see how this is workable. 🤷🏻 

I have a test Lambda function. Used CloudWatch Log Insights to compare before and after.

```
fields @timestamp, @message
| filter @message like /Init Duration/
| parse @message /Init Duration. (?<@init>\d+\.\d+) ms/
| sort @timestamp desc
| limit 20
```

<img width="986" alt="Screenshot 2022-07-24 at 01 19 18" src="https://user-images.githubusercontent.com/1029947/180627104-dbb5260f-6a56-40c6-8e03-7796a847d77a.png">

(The first 50ms coming off the cold start time in the data is due to other optimisations).

**Comment by RichiCoder1** [comment]
@a-h It might be worth cross-posting that to https://github.com/open-telemetry/opentelemetry-lambda/ (which is the underlying repo) as that means the collector itself is taking that long to boot.

**Comment by RichiCoder1** [comment]
Still a very valid issue.

**Comment by ianwremmel** [comment]
In fact, I think it’s gotten worse with the latest releases.

**Comment by adambiggs** [comment]
Any updates? Is this being looked into?

**Comment by mattfysh** [comment]
Achievement unlocked: It's not everyday you stumble upon a Github thread with screenshots of your Slack conversation!

I've been looking into this problem again and it does seem as though ADOT performance on cold starts is still rather poor, but my guess is that it's being caused by the underlying repo https://github.com/open-telemetry/opentelemetry-lambda

If anyone is interested, I'm putting together some automated benchmarks to find the optimal memory setting and implementation when tracing in lambda: https://github.com/mattfysh/lambda-trace-perf

I do find it ironic that we're seeing these bottlenecks. Tracing instrumentations are typically used to measure and monitor application performance, so it does give pause when I see vendors are not fine-tuning their own shared libraries and runtimes required by developers.

**Comment by ianwremmel** [comment]
> I do find it ironic that we're seeing these bottlenecks. 

Agreed. Honestly, I think the real solution here is "stop running the collector in a Lambda layer". I haven't looked into it yet, but I've been noodling on how to run the collector in ECS. I did some manual benchmarking a few weeks ago (I don't have the numbers anymore) and just adding the ADOT layer without even setting the env vars to induce instrumentation adds nearly 800ms. 

Somewhere, I heard that AWS was planning to make xray more OTel native, so hopefully someday it'll be possible to just let the XRay infrastructure do the collection.

**Comment by RichiCoder1** [comment]
I think it's still worth have an "agent" in a layer to take advantage of things like the Telemetry API and handling batch. That said, that agent doesn't necessarily need to be the Otel collector. Maybe something more purpose built would make more sense like a very thin rust agent to collect logs, traces, and batch and forward otel data?

(I promise I'm not mentioning rust cuz it's the cool kid, but rather because real world perf in lambda: https://maxday.github.io/lambda-perf/)

**Comment by ianwremmel** [comment]
yea, there are certainly advantages, but at least for my purposes, the overhead of using any layer seems to be too much of a performance impact to justify using that layer. I've jumped through a lot of goofy hoops to remove multiple layers. From what I can tell, layers only have "good" performance when you hit a scale where coldstarts are an adequately small portion of your traffic that they average their way off of your dashboards.

**Comment by RichiCoder1** [comment]
Bump. As far as I'm aware, this issue hasn't been resolved even with the introduction of the new lambda repo.

**Comment by mattfysh** [comment]
Performance of OTel in Lambda is receiving attention again. If anyone would like to benchmark the current solutions I put together a rudimentary repo here: https://github.com/mattfysh/lambda-trace-perf

I've not looked at it in a while (I've since moved away from serverless because of various cold boot issues, including this one) but if anyone wants to work together on it, I'd be happy to jump back into it

**Comment by deki** [comment]
See also: https://github.com/open-telemetry/opentelemetry-lambda/issues/727

**Comment by jnicholls** [comment]
I'm going to dig into this a little bit this week to try to understand where the startup time is rooted in ADOT's setup. I'll get a baseline with zero components registered and introduce them one by one.

**Comment by RichiCoder1** [comment]
This is still a concern as far as I'm aware.

**Comment by felipenervo** [comment]
Any updates on this issue? We had to remove the layer from our lambdas due to these issues.

**Comment by jnicholls** [comment]
I too removed the layer. It was a nonstarter. An external collector makes more sense for Lambda at this time.

**Comment by ianwremmel** [comment]
@jnicholls would you be able to share anything about your external collector setup? I've been interested in doing something like that, but the pricing has always looked prohibitive.

**Comment by ianwremmel** [comment]
This is still a major issue and it's causing me to run new workloads on fly.io instead of lambda.

**Comment by RangaReddyV** [comment]
any update on this ?
This is a blocking issue for us to use ADOT in lambda
When I compared the performance with ADOT layer and with out ADOT layer for Java 17 runtime, there is a significant amount (~1000ms to ~1200ms) of time taken during cold starts

**Comment by eduardmaurapuig-easyjet** [comment]
We have examples in which lambdas have increased their execution time up to 14x.
<span data-teams="true">
Scenario |  Lambda Layer | OTEL version | Duration | Memory
-- | -- | -- | --: | --:
Nodejs18 w/o ADOT | - | - | 713.20 ms | 108 MB
Nodejs18 w/ ADOT | aws-otel-nodejs-amd64-ver-1-30-1:1 | 0.40.0 | 1123.00 ms | 185 MB
Nodejs22 w/o ADOT | - | - | 792.35 ms | 97 MB
Nodejs22 w/ ADOT | aws-otel-nodejs-amd64-ver-1-30-1:1  | 0.40.0 | **11366.81 ms** | 246 MB
</span>

**Comment by acehand** [comment]
Has there been any update or improvement on this issue. ?

**Comment by mheffner** [comment]
Not to derail the conversation, but we've been working on a lightweight, open-source OpenTelemetry collector written in Rust that might be helpful to others running into cold start latency issues. Our focus has been on minimizing cold starts while keeping the package size and memory usage low, making it a good fit for Lambda environments.

We've packaged it as an AWS Lambda extension to make integration easier. It's still early and has minimal feature support for now, but we're very open to feedback and would love to hear what capabilities might be useful to you.

If you're curious, here’s the repo: https://github.com/streamfold/rotel-lambda-extension

Happy to answer any questions or just chat more about it.


> Unless of course, someone decides to write an OTel collector in Rust. Which would be substantially better imho.

🚀

**Comment by acehand** [comment]
@mheffner will give this a try and get back to you. thank you

**Comment by bilalq** [comment]
@mheffner This looks really interesting! What features are missing for OTLP compatibility? I've avoided OTel due to performance problems, so I'm not really sure how to evaluate whether rotel can handle our needs in its current state.

Let's say I wanted to send to the following destinations:

* Logs
  * Datadog
* Traces
  * Datadog
  * X-Ray
  * Sentry
  * LangFuse (just some spans via a [filterprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md))
* Metrics
  * Clickhouse
  * Datadog (a filtered set of metrics, not all of them)


Is that possible today?

**Comment by mheffner** [comment]
Hey @bilalq , we'd love to help you get rolling with OTel, especially if performance has been an issue. We don't have full support for what you're asking for at the moment, but we're expanding coverage pretty quickly. If you'd like to jump on our [Discord](https://discord.gg/reUqNWTSGC) I'd be happy to discuss priorities and get a better understanding of your environment!

**Comment by RaphaelManke** [comment]
With the latest pricing change aws will charge you for the initial time. 
So it's not only about the inconvenience but now also about money 😅


So I would be really happy to get an official statement from aws on this topic. 

https://aws.amazon.com/blogs/compute/aws-lambda-standardizes-billing-for-init-phase/

**Comment by ianwremmel** [comment]
This remains a problem and the lack of response from AWS is why I'm moving off of AWS in favor of fly.io.

**Comment by vasireddy99** [comment]
ADOT Lambda nodeJS layer is updated to help with optimization and cold starts [please refer here](https://aws-otel.github.io/docs/ReleaseBlogs/aws-distro-for-opentelemetry-lambda-layer-v0.43.0) for more details.
- Specify the OTEL_NODE_ENABLED_INSTRUMENTATIONS environment variable in the lambda to configure the instrumentations needed. By default the layer is set to OTEL_NODE_ENABLED_INSTRUMENTATIONS="aws-sdk,aws-lambda,http" and are loaded automatically by default.
The environment variables should be set to a comma-separated list of the instrumentation package names without the @opentelemetry/instrumentation- prefix.

**Comment by vasireddy99** [comment]
@eduardmaurapuig-easyje do you still see the [issue](https://github.com/aws-observability/aws-otel-lambda/issues/228#issuecomment-2779210178) with latest layer - `ver-1-30-2:1` ?

**Comment by eduardmaurapuig-easyjet** [comment]
I'd have to give it a try, but my schedule is quite tight right now; I can't promise any short update for now. If someone else can do an analysis it would be awesome.

**Comment by ianwremmel** [comment]
for what it's worth, I made an attempt at using the newest layer and my service wouldn't come up. First, I have my own wrapper script around `/opt/otel-handler`:

```bash
#!/usr/bin/env bash

set -euo pipefail

export OTEL_RESOURCE_ATTRIBUTES="${OTEL_RESOURCE_ATTRIBUTES:-}"

if [[ $OTEL_RESOURCE_ATTRIBUTES != *"deployment.environment"* ]]; then
  OTEL_RESOURCE_ATTRIBUTES="deployment.environment=${STAGE_NAME},$OTEL_RESOURCE_ATTRIBUTES"
fi

if [[ $OTEL_RESOURCE_ATTRIBUTES != *"faas.instance"* ]]; then
  OTEL_RESOURCE_ATTRIBUTES="faas.instance=${AWS_LAMBDA_LOG_STREAM_NAME},$OTEL_RESOURCE_ATTRIBUTES"
fi

if [[ $OTEL_RESOURCE_ATTRIBUTES != *"faas.max_memory"* ]]; then
  OTEL_RESOURCE_ATTRIBUTES="faas.max_memory=${AWS_LAMBDA_FUNCTION_MEMORY_SIZE},$OTEL_RESOURCE_ATTRIBUTES"
fi

if [[ $OTEL_RESOURCE_ATTRIBUTES != *"git.commit_sha"* ]]; then
  OTEL_RESOURCE_ATTRIBUTES="git.commit_sha=${SHA},$OTEL_RESOURCE_ATTRIBUTES"
fi

if [[ $OTEL_RESOURCE_ATTRIBUTES != *"git.repository"* ]]; then
  OTEL_RESOURCE_ATTRIBUTES="git.repository=${REPOSITORY_URL},$OTEL_RESOURCE_ATTRIBUTES"
fi

if [[ $OTEL_RESOURCE_ATTRIBUTES != *"service.name"* ]]; then
  SERVICE_NAME=$AWS_LAMBDA_FUNCTION_NAME
  # Remove the trailing random string added by CloudFormation (i.e., everything
  # after the last dash)
  SERVICE_NAME=${SERVICE_NAME%-*}

  # If this is a nested stack, remove the generic "Function" name so we end up
  # using the stack name as the Function name.
  if [[ "$SERVICE_NAME" == *-Function ]]; then
    SERVICE_NAME=${SERVICE_NAME%-*}
  fi

  # Remove the ci-- prefix
  if [[ "$SERVICE_NAME" =~ ci--* ]]; then
    SERVICE_NAME=${SERVICE_NAME/ci--/}
  fi

  # Remove the ci- prefix
  if [[ "$SERVICE_NAME" =~ ci-* ]]; then
    SERVICE_NAME=${SERVICE_NAME/ci-/}
  fi

  # Filter out the sha from the stack name, if present
  while IFS='-' read -ra segment; do
    SERVICE_NAME=''
    for i in "${segment[@]}"; do
      # Remove segments that match the sha
      if [[ "$SHA" == $i* ]]; then
        continue
      fi

      SERVICE_NAME="$SERVICE_NAME-$i"
    done
  done <<< "$SERVICE_NAME"

  # Remove the leading dash that got added by the loop above
  SERVICE_NAME=${SERVICE_NAME#-}

  # At this point, SERVICE_NAME should be
  # <rootStackName>-<partialBranchName>-<functionName>. Given each stanza should
  # be camelcased, it should be safe to simply grab terms 1 and 3 with awk
  # instead of trying to identify the branchName

  SERVICE_NAME=$(echo $SERVICE_NAME | awk -F- '{print $1"-"$3}')

  # At this point, the SERVICE_NAME should be the root stack name followed by
  # either the nested stack name or the function name
  OTEL_RESOURCE_ATTRIBUTES="service.name=${SERVICE_NAME},$OTEL_RESOURCE_ATTRIBUTES"
fi

if [[ $OTEL_RESOURCE_ATTRIBUTES != *"service.version"* ]]; then
  OTEL_RESOURCE_ATTRIBUTES="service.version=${SHA},$OTEL_RESOURCE_ATTRIBUTES"
fi

# Call the ADOT otel handler, which will then call the OpenTelemetry handler
source '/opt/otel-handler'
```

Since it sets `-u` and sources the layer's script, I get a bunch of unbound variable errors. Once I addressed those, I ran into a number of missing dependency errors. I use the following esbuild config in sam and I didn't have time to digging to figure out which otel packages are now included in the layer:

```json
      "Metadata": {
        "BuildMethod": "esbuild",
        "BuildProperties": {
          "EntryPoints": [
            "index"
          ],
          "External": [
            "@aws-sdk/*",
            "@opentelemetry/context-async-hooks",
            "@opentelemetry/core",
            "@opentelemetry/id-generator-aws-xray",
            "@opentelemetry/propagator-aws-xray",
            "@opentelemetry/propagator-b3",
            "@opentelemetry/propagator-jaeger",
            "@opentelemetry/resources",
            "@opentelemetry/sdk-trace-base",
            "@opentelemetry/sdk-trace-node",
            "@opentelemetry/semantic-conventions",
            "@opentelemetry/tracing",
            "aws-sdk"
          ],
          "Minify": false,
          "Sourcemap": true,
          "Target": "es2022"
        }
      },
```

So, basically, there are enough changes here that I don't know when i'll have the time to figure out how to make the new layer work.

**Comment by lalpou** [comment]
Tried the new layer too, managed to get it up and running on one service by simply changing the layer version but for another service I ran into some trouble deploying that I didn't invest time into. For the service that did come up, here are some stats (from a single test run of it on production, so not the most robust numbers):

| otel layer | init (ms) | memory (mb) |
| --- | --- | --- |
| ver-1-30-1:1 | 6859.93 | 455 |
| ver-1-30-2:1 | 5207.91 | 363 | 
| no layer | 1113.03 | 150 |

The new version is the first time init duration has gone below 6s so it had an impact for sure. Not sure if the memory use being lower was also a consistent thing, but saw <6s init pretty consistently

**Comment by bilalq** [comment]
X-ray being deprecated now makes this issue even more critical. It's crazy to deprecate it with such a huge performance penalty on Otel.

**Comment by ianwremmel** [comment]
This is the first I'm hearing about XRay's deprecation, but after a quick read, I'm betting the right solution is give up on ADOT and just use the language-appropriate SDKs to write directly to the [OTEL Endpoints](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-OTLPEndpoint.html)

**Comment by fitz-vivodyne** [comment]
To clarify, the X-Ray SDKs and daemon are being deprecated as described in this blog post: https://aws.amazon.com/blogs/mt/aws-x-ray-sdks-daemon-migration-to-opentelemetry/
